docs/design/batch-request.md

```diff
@@ -0,0 +1,260 @@
+
+## Configuration
+1. **BATCH_REQUESTS_ENABLED:** Flag to enable `batch requests`, it can be `true` or `false`, default value is `false`
+2. **BATCH_REQUESTS_MAX_SIZE:** If the batch request is enabled, this is the maximum number of requests allowed in a batch request, default value is `100`.
+3. **WS_BATCH_REQUESTS_ENABLED:** Flag to enable `batch requests` on the websocket, it can be `true` or `false`, default value is `false`
+4. **WS_BATCH_REQUESTS_MAX_SIZE:** If the batch request is enabled, this is the maximum number of requests allowed in a batch request, default value is `20`.
+
+
+## Implementation
+
+* Current single requests should stay the same as they are currently.
+
+* All batch requests should return a `200` response code, even if there are errors on the individual requests. except in case that fails the validations for the batch request as a whole or common validations.
+* Batch requests should be implemented on top of the current implementation, refector into common flow and common code as much as possible.
+
+**Common Flow for both Batch and Single Requests:**
+1. Validate the request is a valid json
+2. Validate the Method is supported (POST only)
+3. Validate Authorization Token
+
+**Validations per Each individual request on a batch or as a single request:**
+1. Validate JSON-RPC 2.0 Specification
+2. Validate Method is supported
+3. Validate and increase the request count for the rate limit
+
+**Validations for the Batch Request as a whole:**
+1. Validate that the request is an array
+2. Validate Batch Request is enabled
+3. Validate Batch Request size is within the limit
+
+
+
+#### Batch Request examples
+
+The request object is a JSON array with one or more transactions, a unique `id` should be provided for each request (however due to the experimental feature that allows to send a request without `id` this is not mandatory, and answers will be provided in the same order as received), and all other fields as defined in the [JSON-RPC 2.0 Specification](https://www.jsonrpc.org/specification) are also needed to comform to the standard.
+
+```json
+[
+    {
+        "jsonrpc": "2.0",
+        "id": 1,
+        "method": "eth_chainId",
+        "params": []
+    },
+    {
+        "id": "1",
+        "jsonrpc": "2.0",
+        "method": "eth_getBalance",
+        "params": [
+            "0x67D8d32E9Bf1a9968a5ff53B87d777Aa8EBBEe69",
+            "latest"
+        ]
+    },
+    {
+        "jsonrpc": "2.0",
+        "method": "eth_blockNumber",
+        "params": [],
+        "id": 1
+    }
+]
+```
+
+#### Batch Response example
+
+The response object is a JSON array with each result from the request array. Each response object has the same `id` that corresponds to the request object.
+
+```json
+[
+    {
+        "result": "0x12a",
+        "jsonrpc": "2.0",
+        "id": 1
+    },
+    {
+        "result": "0x21e19e0c9bab2400000",
+        "jsonrpc": "2.0",
+        "id": "1"
```

Comment:
set the ids to unique.
Actually is this something we need to validate as part of the batch logic also?

Reply:
it might be a good consideration, although Alchemy or infura let you send repeated ids and they both return the items in the same order, and since `id` might be optional in some cases. I did not add that validation.

Reply:
Okay let's leave validation out, please note this in the design doc.